packages/core/logging/core-logging-common-internal/src/logger.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import {
+  Appender,
+  LogLevel,
+  LogRecord,
+  LoggerFactory,
+  LogMeta,
+  Logger,
+  LogLevelId,
+} from '@kbn/logging';
+
+/**
+ * @internal
+ */
+export type CreateLogRecordFn = <Meta extends LogMeta>(
+  level: LogLevel,
+  errorOrMessage: string | Error,
+  meta?: Meta
+) => LogRecord;
+
+/**
+ * A basic, abstract logger implementation that delegates the create of log records to the child's createLogRecord function.
+ * @internal
+ */
+export abstract class AbstractLogger implements Logger {
+  constructor(
+    protected readonly context: string,
+    protected readonly level: LogLevel,
+    protected readonly appenders: Appender[],
+    protected readonly factory: LoggerFactory
+  ) {}
+
+  protected abstract createLogRecord<Meta extends LogMeta>(
```

Comment:
TIL (again): `protected` vs `private`. I had to read up on `protected` vs `private` class methods. The difference is subtle.

Reply:
Yea. I almost never use `protected` in typescript, but for abstract class, the abstract method definitions must be declared as protected, else typescript complains.